fix(gitlab): smart truncate comments

amishk783 opened the pull request:
<!-- If this is your first pull request: sign the CLA with this GitHub app: https://cla-assistant.io/renovatebot/renovate -->
<!-- Make sure the `Allow edits and access to secrets by maintainers` checkbox is checked on this pull request. -->
<!-- Please read https://github.com/renovatebot/renovate/blob/main/.github/contributing.md before you create your pull request.-->

## Changes

<!-- Describe what behavior is changed by this PR. -->

## Context
- closes #30551
<!-- Describe why you're making these changes if it's not already explained in a corresponding issue. -->
<!-- If you're closing an existing issue with this pull request, use the keyword Closes #issue_number. -->
<!-- If you're referencing an issue with this pull request, put it in a Markdown list like this: - #issue_number. -->

## Documentation (please check one with an [x])

- [ ] I have updated the documentation, or
- [x] No documentation update is required

## How I've tested my work (please select one)

I have verified these changes via:

- [ ] Code inspection only, or
- [ ] Newly added/modified unit tests, or
- [ ] No unit tests but ran on a real repository, or
- [ ] Both unit tests + ran on a real repository

<!-- Do you have any suggestions about this PR template? Edit it here: https://github.com/renovatebot/renovate/edit/main/.github/pull_request_template.md -->

<!-- Please do not force push to your PR's branch after you have created your PR, as doing so forces us to review the whole PR again. This makes it harder for us to review your work because we don't know what has changed. -->
<!-- PRs will always be squashed by us when we merge your work. Commit as many times as you need in this branch. -->

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/renovatebot/renovate?pullRequest=30580) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/renovatebot/renovate?pullRequest=30580) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/renovatebot/renovate?pullRequest=30580) it.</sub>

RahulGautamSingh reviewed (changes requested):
Trunacate the body content before we compute commentNeedsUpdating. Else, we will compare full body content with comment body we fetch from api. And, the fetched body will be truncated whereas full body (from fn param) won't. 

This will cause the comment to be updated each time.

amishk783 commented:
Hi Rahul, Thanks for the guidance. I’ve moved smartTruncate(body, maxBodyLength()) before the commentNeedsUpdating check in the if-else case and will push the changes shortly.

henry-fn commented:
Thanks @amishk783 for the fix, anything blocking this being merged in?

viceice commented:
> Thanks @amishk783 for the fix, anything blocking this being merged in?

required changes are not pushed. also CI falls, so merge isn't an option.

amishk783 commented:
> > Thanks @amishk783 for the fix, anything blocking this being merged in?
> 
> required changes are not pushed. also CI falls, so merge isn't an option.

Hey, sorry for the delay. it completely slipped my mind. I've pushed the changes and addressed the feedback. Now truncating the body before checking commentNeedsUpdating

henry-fn commented:
> > > Thanks @amishk783 for the fix, anything blocking this being merged in?
> > 
> > 
> > required changes are not pushed. also CI falls, so merge isn't an option.
> 
> Hey, sorry for the delay. it completely slipped my mind. I've pushed the changes and addressed the feedback. Now truncating the body before checking commentNeedsUpdating

Just in case you missed it, I think there is still one linting test failing https://github.com/renovatebot/renovate/actions/runs/10960183807/job/30434310107?pr=30580

henry-fn commented:
@amishk783 poke on formatting here?

henry-fn commented:
Anything blocking merging here?

rarkins commented:
@viceice please review

henry-fn commented:
Checking in again here

viceice commented on lib/modules/platform/gitlab/index.ts:
```diff
@@ -1400,13 +1401,15 @@ export async function ensureComment({
   } else {
     logger.debug(`Ensuring content-only comment in #${number}`);
     body = `${sanitizedContent}`;
+    body = smartTruncate(body, maxBodyLength());
```
```suggestion
    body = smartTruncate(`${sanitizedContent}`, maxBodyLength());
```

viceice commented on lib/modules/platform/gitlab/index.ts:
```diff
@@ -1391,6 +1391,7 @@ export async function ensureComment({
     body = body
       .replace(regEx(/Pull Request/g), 'Merge Request')
       .replace(regEx(/PR/g), 'MR');
+    body = smartTruncate(body, maxBodyLength());
```
```suggestion
     body = smartTruncate(body
       .replace(regEx(/Pull Request/g), 'Merge Request')
       .replace(regEx(/PR/g), maxBodyLength());
```
needs prettifying

jwilliams-fn commented:
@amishk783 poking here for _hopefully_ the final time 😅

RahulGautamSingh commented:
Replacement PR: https://github.com/renovatebot/renovate/pull/33348
